### PR TITLE
fix(weak-delegate): except appdelegate with SwiftUI's @NSApplicationDelegateAdaptor wrapper

### LINKED ISSUE #3598: `weak_delegate`: @NSApplicationDelegateAdaptor should be excepted
### New Issue Checklist

- [x] Updated SwiftLint to the latest version
- [x] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

### New rule request

This is a complement to #3456:

The weak delegate rule fires on the following declaration, which is not appropriate:

@NSApplicationDelegateAdaptor(AppDelegate.self) var appDelegate

I will create a PR follow up PR, complementing the exception for these SwiftUI specified wrappers.

### LINKED ISSUE #3611: weak_delegate: @WKExtensionDelegateAdaptor should be non-triggering
### New Issue Checklist

- [x] Updated SwiftLint to the latest version
- [x] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

### Describe the bug

WKExtensionDelegateAdaptor triggers weak_delegate but should not.

##### Complete output when running SwiftLint, including the stack trace and command used

```bash
% swiftlint
Linting Swift files in current working directory
...
Linting 'Application.swift' (12/14)
...
/Users/username/Developer/Applications/ApplicationName/Sources/Application/Application.swift:19:57: warning: Weak Delegate Violation: Delegates should be weak to avoid reference cycles. (weak_delegate)
Done linting! Found 1 violation, 0 serious in 14 files.
```

### Environment

* SwiftLint version (run `swiftlint version` to be sure)? 0.43.1
* Installation method used (Homebrew, CocoaPods, building from source, etc)? pkg install
* Paste your configuration file:

```yml
disabled_rules:
  - line_length

trailing_whitespace:
  ignores_empty_lines: true

```

* Are you using [nested configurations](https://github.com/realm/SwiftLint#nested-configurations)? No
* Which Xcode version are you using (check `xcodebuild -version`)? Xcode 12.5 Build version 12E262
* Do you have a sample that shows the issue? See example below

```swift
import SwiftUI

@main
struct Application: App {

    #if os(iOS)

    @UIApplicationDelegateAdaptor(ApplicationDelegate.self) var applicationDelegate

    #elseif os(watchOS)

    @WKExtensionDelegateAdaptor(ExtensionDelegate.self) var extensionDelegate

    #endif

    let persistenceController = PersistenceController.shared

    @SceneBuilder var body: some Scene {
        WindowGroup {
            ContentView()
                .environment(\.managedObjectContext, persistenceController.container.viewContext)
        }

        #if os(watchOS)

        WKNotificationScene(controller: NotificationController.self, category: "myCategory")

        #endif
    }
}
```

### PR DESCRIPTION
**Pull request by tillhainbach**
should fix #3598

### Edit:
should also fix  #3611

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -11,7 +11,9 @@
 
 #### Enhancements
 
-* None.
+* Don't report `@NSApplicationDelegateAdaptor` statements in
```
Comment (marcelofabri) [inline CHANGELOG.md:14]: I think this is missing some words :P

**Review by marcelofabri — APPROVED** [review]

File: Source/SwiftLintFramework/Rules/Lint/WeakDelegateRule.swift [hunk 2]
```diff
@@ -81,7 +82,7 @@ public struct WeakDelegateRule: ASTRule, SubstitutionCorrectableASTRule, Configu
                 let offset = attribute.offset,
                 let length = attribute.length,
                 let value = file.stringView.substringWithByteRange(ByteRange(location: offset, length: length)),
-                value.hasPrefix("@UIApplicationDelegateAdaptor") {
+                (value.hasPrefix("@UIApplicationDelegateAdaptor") || value.hasPrefix("@NSApplicationDelegateAdaptor")) {
```
Comment (marcelofabri) [inline Source/SwiftLintFramework/Rules/Lint/WeakDelegateRule.swift:85]: can you please update this to handle `WKExtensionDelegateAdaptor` and fix https://github.com/realm/SwiftLint/issues/3611 as well? 🙏

File: CHANGELOG.md (hunk 1 above)
Comment (tillhainbach) [reply CHANGELOG.md]: ups 🙈

File: Source/SwiftLintFramework/Rules/Lint/WeakDelegateRule.swift (hunk 2 above)
Comment (tillhainbach) [reply Source/SwiftLintFramework/Rules/Lint/WeakDelegateRule.swift]: Done!

**Comment by SimplyDanny** [comment]
@tillhainbach: I update the PR fixing the conflicts. It will be merged as soon as the checks report success.

**Pull request merged**